### PR TITLE
Implement password encryption using an RSA public key

### PR DESCRIPTION
**Pull request by KarboniteKream**
## Description
This PR is a follow-up to #358, to implement password encryption over unsafe connection using an RSA public key. This is required for `priority #2` as described on [this page](https://dev.mysql.com/doc/dev/mysql-server/latest/page_caching_sha2_authentication_exchanges.html).

## Detailed changes
* Add `rsaPublicKey` property to `Configuration`
* Add authentication state to `MySQLConnection`, and other configuration to `HandshakeResponse`, so they can be available during authentication phases
* Implement password encryption using a public key in `Sha256PasswordAuthentication`
* Provide custom public/private key pairs to test containers

### CODE REVIEW

**Review by oshai — COMMENTED** [review]
Thank you very much for this effort! It's appreciated!
Other than the link to mysql I think it will be very hard to follow how auth works here in the driver.
And it looks like you pretty much revisited all auth methods here?
What do you think about adding an additional readme with impl details? So it will be more clear later, and easier to follow in the future.

File: mysql-async/src/main/java/com/github/jasync/sql/db/mysql/MySQLConnection.kt [hunk 1]
```diff
@@ -298,7 +304,8 @@ class MySQLConnection @JvmOverloads constructor(
             message.authenticationMethod,
             database = configuration.database,
             password = configuration.password,
-            appName = configuration.applicationName
+            appName = configuration.applicationName,
+            configuration = configuration,
```
[the remark this answers is not in the document]
Comment (KarboniteKream) [reply mysql-async/src/main/java/com/github/jasync/sql/db/mysql/MySQLConnection.kt:308]: I need it on `Sha256Password`, to behave differently based on SSL mode. But I can just pass `SSLConfiguration` instead.

[the remark this answers is not in the document]
Comment (oshai) [reply mysql-async/src/main/java/com/github/jasync/sql/db/mysql/MySQLConnection.kt:308]: SSLConfiguration is better.

**Comment by KarboniteKream** [comment]
> What do you think about adding an additional readme with impl details?

Will do! I agree it's difficult to follow, and I was thinking of extracting the authentication logic into a separate class to handle all of this. For now I'll make README explaining how it all works, and sometime later work on a prototype to hopefully simplify the whole flow.

File: db-async-common/src/main/java/com/github/jasync/sql/db/Configuration.kt [hunk 2]
```diff
@@ -72,6 +76,10 @@ class Configuration @JvmOverloads constructor(
                     "Please change eventLoopGroup configuration."
             }
         }
+
+        if (rsaPublicKey != null && rsaPublicKey.notExists()) {
```
[the remark this answers is not in the document]
Comment (KarboniteKream) [reply db-async-common/src/main/java/com/github/jasync/sql/db/Configuration.kt:80]: Ah, I didn't consider this, thank you. I'll see how to approach this, perhaps I can just remove it completely and only keep the try-catch when reading the file.

File: mysql-async/src/main/java/com/github/jasync/sql/db/mysql/MySQLConnection.kt (hunk 1 above)
[the remark this answers is not in the document]
Comment (KarboniteKream) [reply mysql-async/src/main/java/com/github/jasync/sql/db/mysql/MySQLConnection.kt:308]: I actually need both `configuration.ssl` and `configuration.rsaPublicKey`. Would you prefer that I add both parameters to `generateAuthentication()`, or should I pass `password` and `configuration`? In the latter case, I would not use `configuration.password`.

[the remark this answers is not in the document]
Comment (oshai) [reply mysql-async/src/main/java/com/github/jasync/sql/db/mysql/MySQLConnection.kt:308]: I think it's better to pass what actually needed and not the configuration object.

**Comment by KarboniteKream** [comment]
I've addressed the comments. Please take another look.

File: db-async-common/src/main/java/com/github/jasync/sql/db/Configuration.kt [hunk 3]
```diff
@@ -63,7 +65,8 @@ class Configuration @JvmOverloads constructor(
     val executionContext: Executor = ExecutorServiceUtils.CommonPool,
     val currentSchema: String? = null,
     val socketPath: String? = null,
-    val credentialsProvider: CredentialsProvider? = null
+    val credentialsProvider: CredentialsProvider? = null,
+    val rsaPublicKey: Path? = null,
```
Comment (oshai) [inline db-async-common/src/main/java/com/github/jasync/sql/db/Configuration.kt:69]: sorry for asking just now (didn't thought about it before). shouldn't rsaPublicKey be part of  SSLConfiguration? I mean is there a usage for the key without ssl?

File: mysql-async/src/main/java/com/github/jasync/sql/db/mysql/encoder/auth/README.md [hunk 4]
```diff
@@ -0,0 +1,60 @@
+# Authentication methods
```
Comment (oshai) [inline mysql-async/src/main/java/com/github/jasync/sql/db/mysql/encoder/auth/README.md:1]: Awesome!

File: db-async-common/src/main/java/com/github/jasync/sql/db/Configuration.kt (hunk 3 above)
Comment (KarboniteKream) [reply db-async-common/src/main/java/com/github/jasync/sql/db/Configuration.kt:69]: I had it there initially, but I think it makes more sense to be separate. Since you can decide not to use SSL (use unsafe connection), in which case you need RSA to encrypt the password. So the RSA public key does not imply SSL or the other way around. But if you prefer, I can move it to `SSLConfiguration`

Comment (oshai) [reply db-async-common/src/main/java/com/github/jasync/sql/db/Configuration.kt:69]: if they are independent, I am ok with leaving it as is.

**Comment by oshai** [comment]
Let me know if it's ready to merge or anything else you'd like to change before I merge.

**Comment by KarboniteKream** [comment]
It's ready for merge, thanks!

**Comment by oshai** [comment]
Thanks you for the PR!
